### PR TITLE
Events in week view

### PR DESCRIPTION
**Pull request by RokasMesk**
now the events, can be displayed in the week view, u can click on them to get info about them.
NOTE:
i installed extension that formats the document, so a lot of changes are spaces and tabs.

### CODE REVIEW

File: css/styles.css [hunk 1]
```diff
@@ -1,148 +1,161 @@
...
-    grid-template-rows:auto 1fr;
-    grid-template-columns: 200px 1fr;
-    height: 100vh;
-    margin: 0;
-    font-family: Arial, sans-serif;
-    overflow: none;
-    overscroll-behavior: none;
+  display: grid;
+  grid-template-rows: auto 1fr;
+  grid-template-columns: 400px 1fr;
```
Comment (mildasuziedelyte) [inline css/styles.css]: not sure why this change, since layout now seems a bit broken for me:
image

File: js/main.js [hunk 2]
```diff
@@ -1,12 +1,15 @@
...
+import { loadEventsForCurrentWeek } from "./events.js";
+
 const currentDate = new Date();
+
 function init() {
-    renderCalendarCells();
-    initModalEventListeners();
-    renderWeekHeader(currentDate);
-    addEventListenersToHeaderButtons();
+  renderCalendarCells(() => loadEventsForCurrentWeek(currentDate));
```
Comment (mildasuziedelyte) [inline js/main.js]: as we spoke, calendar itself could handle events loading based on the date.

```suggestion
  renderCalendarCells(currentDate);
```

File: js/calendar.js [hunk 3]
```diff
@@ -1,43 +1,39 @@
...
-    return new Date(date.setDate(diff));
-}
-function isToday(date) {
-    const today = new Date();
-    return date.getDate() === today.getDate() &&
-           date.getMonth() === today.getMonth() &&
-           date.getFullYear() === today.getFullYear();
+import { getStartOfWeek, isToday } from "./utils.js";
+
+export let startDayOfWeek = new Date();
```
Comment (mildasuziedelyte) [inline js/calendar.js]: I don't see it used in other file so no export needed.
also, since value is current date, naming is not correct. if it is current date I do not think you should change it, so it should be const.

In case you need variable where you save same value that you change in loop, it should be undefined at first, before you assign value. 
then if I understand correctly what you use this variable for it could be named like weekDay or smth like that : )

```suggestion
const currentDate = new Date();
```

File: js/utils.js [hunk 4]
```diff
@@ -0,0 +1,13 @@
+export function getStartOfWeek(date) {
```
Comment (mildasuziedelyte) [inline js/utils.js]:
```suggestion
export function getFirstDayOfTheWeek(date) {
```

File: js/calendar.js [hunk 5]
```diff
@@ -1,43 +1,39 @@
...
-        dayHeader.className = 'day';
-        dayHeader.innerHTML = `<span class='day-name'>${dayName}</span><span class="day-number">${day.getDate()}</span>`;
-        if (isToday(day)) {
-            dayHeader.classList.add('current-day');
-        }
-        weekHeader.appendChild(dayHeader);
+export function renderWeekHeader(dateToRender) {
+  const weekHeader = document.querySelector(".week-header");
+  weekHeader.innerHTML = "";
+  const startOfWeek = getStartOfWeek(dateToRender);
```
Comment (mildasuziedelyte) [inline js/calendar.js]:
```suggestion
  const firstDayOfTheWeek = getFirstDayOfTheWeek(dateToRender);
```

File: js/calendar.js [hunk 6]
```diff
@@ -1,43 +1,39 @@
...
+    const daysInWeek = 7;
+    for (let i = 0; i < cellsInOneColumn * daysInWeek; i++) {
+      const cell = document.createElement("div");
+      cell.className = "cell";
+      calendarCells.appendChild(cell);
+    }
+    if (typeof loadEventsCallback === "function") {
+      loadEventsCallback();
+    }
+  });
```
Comment (mildasuziedelyte) [inline js/calendar.js]: let's start splitting to smaller functions, so that one function does pretty much one thing

```suggestion
export function renderMainCalendar(date) {
  document.addEventListener("DOMContentLoaded", () => {
    // we could also move week header rendering here as well
    renderCalendarCells(date)
    loadWeekEvents(date)
  });
```

why did you need check typeof loadEventsCallback === "function" ?

File: js/calendar.js [hunk 7]
```diff
@@ -1,43 +1,39 @@
...
+export function renderWeekHeader(dateToRender) {
+  const weekHeader = document.querySelector(".week-header");
+  weekHeader.innerHTML = "";
+  const startOfWeek = getStartOfWeek(dateToRender);
+  const wholeWeekHeader = document.createElement("div");
+  wholeWeekHeader.className = "day";
+  for (let i = 0; i < 7; i++) {
+    const day = new Date(startOfWeek);
+    day.setDate(day.getDate() + i);
+    if (i === 0) startDayOfWeek = day;
```
Comment (mildasuziedelyte) [inline js/calendar.js]: there are two variables:
- startOfWeek
- startDayOfWeek
that from the naming seems to have same value. 
why do you need two?
maybe the naming could be improved?

this needs to be tidied up, cause it is hard to read and understand what is happening..

File: js/events.js [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+    const endDayIndex = startDayIndex + totalDays - 1;
+    eventBar.style.gridColumn = `${startDayIndex + 1} / ${endDayIndex + 2}`;
+    weekHeader.appendChild(eventBar);
+    eventBar.addEventListener("click", () => showEventDetails(event));
+  }
+  eventElement.addEventListener("click", () => showEventDetails(event));
+}
+
+export function saveEventToLocalStorage(event) {
+  let events = JSON.parse(localStorage.getItem("calendarEvents")) || [];
```
Comment (mildasuziedelyte) [inline js/events.js]: if you want events to default to some value if localStorage does not have any events saved, it is better to use nullish coalescing operator
https://javascript.info/nullish-coalescing-operator

File: js/events.js [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+  events.forEach((event) => {
+    const eventStartDateTime = new Date(event.startDateTime);
+    if (eventStartDateTime >= startOfWeek && eventStartDateTime <= endOfWeek) {
+      console.log("true");
+      createEventElement(event);
+    }
+  });
+}
+
+export function getStartOfWeek(date) {
```
Comment (mildasuziedelyte) [inline js/events.js]: move to utils file

File: js/events.js [hunk 10]
```diff
@@ -0,0 +1,103 @@
+export function createEventElement(event) {
+  const eventElement = document.createElement("div");
+  eventElement.className = "calendar-event";
+  eventElement.innerHTML = `<strong>${event.title}</strong><br>${event.description}`;
+  const startDateTime = new Date(event.startDateTime);
+  const endDateTime = new Date(event.endDateTime);
+  let startDay = startDateTime.getDay();
+  let dayOfWeek = startDateTime.getDay();
```
Comment (mildasuziedelyte) [inline js/events.js]: why two of the same?

File: js/events.js [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+  const startHour = startDateTime.getHours() + startDateTime.getMinutes() / 60;
+  const endHour = endDateTime.getHours() + endDateTime.getMinutes() / 60;
+  const totalDays = Math.ceil(
+    (endDateTime - startDateTime) / (1000 * 60 * 60 * 24)
+  );
+  console.log("total days " + totalDays);
```
Comment (mildasuziedelyte) [inline js/events.js]:
```suggestion
```
let's not commit console.logs

File: js/events.js [hunk 12]
```diff
@@ -0,0 +1,103 @@
...
+  events.push(event);
+  localStorage.setItem("calendarEvents", JSON.stringify(events));
+}
+
+export function loadEventsFromLocalStorage() {
+  let events = JSON.parse(localStorage.getItem("calendarEvents")) || [];
+  events.forEach((event) => {
+    createEventElement(event);
+  });
+}
```
Comment (mildasuziedelyte) [inline js/events.js]: this function does too many things.
one function should load events, ie. return events from local storage
another function should render events

File: js/events.js [hunk 13]
```diff
@@ -0,0 +1,103 @@
...
+    eventBar.id = "multipleDayEvent";
+    eventBar.innerHTML = `<strong>${event.title}</strong>`;
+    const startDayIndex = startDay === 0 ? 7 : startDay;
+    const endDayIndex = startDayIndex + totalDays - 1;
+    eventBar.style.gridColumn = `${startDayIndex + 1} / ${endDayIndex + 2}`;
+    weekHeader.appendChild(eventBar);
+    eventBar.addEventListener("click", () => showEventDetails(event));
+  }
+  eventElement.addEventListener("click", () => showEventDetails(event));
+}
```
Comment (mildasuziedelyte) [inline js/events.js]: if this function handles both short and multi day events, let's split it to smaller functions for better readability, like:
renderEvents = (events) => {
   // split events to get short events and multi day events
   // render short events
   // render long events
}

by doing this there should be less ifs, also try to move code to utils when possible

File: index.html [hunk 14]
```diff
@@ -1,105 +1,188 @@
...
-            <div class="timestamp">7:00 PM</div>
-            <div class="timestamp">8:00 PM</div>
-            <div class="timestamp">9:00 PM</div>
-            <div class="timestamp">10:00 PM</div>
-            <div class="timestamp">11:00 PM</div>
-        </div>
-        <div class="calendar-cells" id="calendarCells">
-        </div>
+      <div class="week-header"></div>
+      <div class="multi-day-event-placeholder">
```
Comment (mildasuziedelyte) [inline index.html]:
```suggestion
      <div class="multi-day-events-container">
```

File: js/main.js [hunk 15]
```diff
@@ -1,12 +1,24 @@
...
+import { renderAside } from './aside.js';
+import { renderCalendarCells } from './calendar.js';
+import { initModal } from './modal.js';
+
+const init = () => {
+  let currentDate = new Date();
+  const { openModal } = initModal();
+
+  const updateDate = (newDate) => {
+    if (newDate.toDateString() === currentDate.toDateString()) {
```
Comment (mildasuziedelyte) [inline js/main.js]: I don't think toDateString is correct way to compare if dates are same, since it is more meant for date representation in a string format.

Since I guess you want to compare just year, month and day, you should go with getFullYear, getMonth and getDate methods.

File: js/header.js [hunk 16]
```diff
@@ -1,34 +1,50 @@
...
-function updateCurrentYearAndMonth(){
-    const currentMonthAndDaySpan = document.getElementById('currentMonthAndDay');
-    const options = { year: 'numeric', month: 'short' };
-    currentMonthAndDaySpan.textContent = currentDate.toLocaleDateString('en-US', options);
-}
-function switchWeeks(days) {
-    currentDate.setDate(currentDate.getDate() + days);
-    updateCurrentYearAndMonth();
-    renderWeekHeader(currentDate);
+export const renderWeekHeader = (date) => {
```
Comment (mildasuziedelyte) [inline js/header.js]: this should move to calendar.js, since it is not part of the header actually.

File: js/header.js [hunk 17]
```diff
@@ -1,34 +1,50 @@
...
+    currentMonthAndDaySpan.textContent = date.toLocaleDateString("en-US", options);
+  };
 
-function eventListenerForSwitchingWeeksButton() {
-    document.getElementById('todayButton').addEventListener('click', () => {
-        currentDate = new Date();
-        switchWeeks(0);
-    });
+  const switchWeeks = (days) => {
+    const newDate = new Date(date);
```
Comment (mildasuziedelyte) [inline js/header.js:9]: do you really need this? since date that comes as renderHeader param is already Date type

File: js/events.js [hunk 18]
```diff
@@ -0,0 +1,70 @@
...
+  
+  const events = getEventsFromLocalStorage();
+  
+  events.forEach((event) => {
+    const eventStartDateTime = new Date(event.startDateTime);
+    const eventEndDateTime = new Date(event.endDateTime);
+    const totalDays = Math.ceil((eventEndDateTime - eventStartDateTime) / MILLISECONDS);
+    
+    if (eventStartDateTime <= endOfWeek && eventEndDateTime >= startOfWeek) {
+      if (totalDays <= 1) {
```
Comment (mildasuziedelyte) [inline js/events.js]: move conditions to variables:

```suggestion
const shouldRenderEvent = eventStartDateTime <= endOfWeek && eventEndDateTime >= startOfWeek
const isShortEvent = totalDays <= 1

    if (shouldRenderEvent) {
      if (isShortEvent) {
```

File: js/calendar.js [hunk 19]
```diff
@@ -1,43 +1,17 @@
...
+  calendarCells.innerHTML = ''; 
+  const cellsInOneColumn = 19;
+  const daysInWeek = 7;
+  for (let i = 0; i < cellsInOneColumn * daysInWeek; i++) {
+    const cell = document.createElement("div");
+    cell.className = "cell";
+    cell.addEventListener('click', () => openModal(date)); 
+    calendarCells.appendChild(cell);
+  }
+  renderWeekHeader(date);
```
Comment (mildasuziedelyte) [inline js/calendar.js:24]: since you have renderWeekHeader here (which I think is better place), do you still need to renderWeekHeader in header.js?

File: js/utils.js [hunk 20]
```diff
@@ -0,0 +1,57 @@
...
+    date.getMonth() === today.getMonth() &&
+    date.getFullYear() === today.getFullYear()
+  );
+}
+
+export function getEventsFromLocalStorage() {
+  let events = JSON.parse(localStorage.getItem("calendarEvents")) || [];
+  return events;
+}
+
```
Comment (mildasuziedelyte) [inline js/utils.js]: these kind of functions that get or set something from database (in this case localStorage) we usually call services and they live in their own services.js file.

why it caught my attention now is because get from local storage and save in local storage are in different files. should be in same : )

File: js/modal.js [hunk 21]
```diff
@@ -1,38 +1,76 @@
...
+  };
+
+  const closeEventCreationModal = () => {
     modal.style.display = "none";
-}
-function validateEventForm(event) {
+  };
+
+  const openModal = (date) => {
+    //kai reikes jeigu reikes
```
Comment (mildasuziedelyte) [inline js/modal.js]: : )
I think we already need it.
I am not sure if you need both openEventCreationModal and openModal though.

File: js/modal.js [hunk 22]
```diff
@@ -1,38 +1,76 @@
...
     const endTime = document.getElementById("endTime").value;
+    const eventDescription = document.getElementById("eventDescription").value;
+
     const startDateTime = new Date(`${startDate}T${startTime}`);
     const endDateTime = new Date(`${endDate}T${endTime}`);
 
     if (startDateTime > endDateTime) {
-        event.preventDefault();
-        alert("End date and time must be after start date and time");
+      alert("End date and time must be after start date and time");
```
Comment (mildasuziedelyte) [inline js/modal.js]: would be nicer to display error below the input. alert is not actually used in real life

File: js/header.js (hunk 17 above)
Comment (RokasMesk) [reply js/header.js:9]: In my opinion I do need this, I have to create a new object, because Date object is mutable (changing date object in switchWeeks results in general date change to the whole renderHeader). I tested, it doesn't work without creation of copy, doesn't render.

File: js/calendar.js [hunk 23]
```diff
@@ -1,17 +1,46 @@
...
   for (let i = 0; i < cellsInOneColumn * daysInWeek; i++) {
     const cell = document.createElement("div");
     cell.className = "cell";
-    cell.addEventListener('click', () => openModal(date)); 
+    const day = i % daysInWeek;
+    const hourIndex = Math.floor(i / daysInWeek); 
+    const cellDate = new Date(startOfWeek);
+    cellDate.setDate(cellDate.getDate() + day);
+    cellDate.setHours(5 + hourIndex); 
+    cell.addEventListener('click', () => openModal(cellDate));
```
Comment (mildasuziedelyte) [inline js/calendar.js:20]: could be moved to separate function addCellClickEventListener

**Review by mildasuziedelyte — APPROVED** [review]
